src/Symfony/Component/Polyfill/Normalizer.php

```diff
@@ -0,0 +1,292 @@
+                // ASCII chars
+
+                if ($tail) {
+                    $lastUchr .= $tail;
+                    $tail = '';
+                }
+
+                if ($j = strspn($s, $ASCII, $i + 1)) {
+                    $lastUchr .= substr($s, $i, $j);
+                    $i += $j;
+                }
+
+                $result .= $lastUchr;
+                $lastUchr = $s[$i];
+                ++$i;
+            } else {
+                $ulen = $ulenMask[$s[$i] & "\xF0"];
+                $uchr = substr($s, $i, $ulen);
+
+                if ($lastUchr < "\xE1\x84\x80" || "\xE1\x84\x92" < $lastUchr
+                    ||   $uchr < "\xE1\x85\xA1" || "\xE1\x85\xB5" < $uchr
+                    || $lastUcls) {
+                    // Table lookup and combining chars composition
+
+                    $ucls = isset($combClass[$uchr]) ? $combClass[$uchr] : 0;
+
+                    if (isset($compMap[$lastUchr.$uchr]) && (!$lastUcls || $lastUcls < $ucls)) {
+                        $lastUchr = $compMap[$lastUchr.$uchr];
+                    } elseif ($lastUcls = $ucls) {
+                        $tail .= $uchr;
+                    } else {
+                        if ($tail) {
+                            $lastUchr .= $tail;
+                            $tail = '';
+                        }
+
+                        $result .= $lastUchr;
+                        $lastUchr = $uchr;
+                    }
+                } else {
+                    // Hangul chars
+
+                    $L = ord($lastUchr[2]) - 0x80;
+                    $V = ord($uchr[2]) - 0xA1;
+                    $T = 0;
+
+                    $uchr = substr($s, $i + $ulen, 3);
+
+                    if ("\xE1\x86\xA7" <= $uchr && $uchr <= "\xE1\x87\x82") {
+                        $T = ord($uchr[2]) - 0xA7;
+                        0 > $T && $T += 0x40;
+                        $ulen += 3;
+                    }
+
+                    $L = 0xAC00 + ($L * 21 + $V) * 28 + $T;
+                    $lastUchr = chr(0xE0 | $L >> 12).chr(0x80 | $L >> 6 & 0x3F).chr(0x80 | $L & 0x3F);
+                }
+
+                $i += $ulen;
+            }
+        }
+
+        return $result.$lastUchr.$tail;
+    }
+
+    private static function decompose($s, $c)
+    {
+        $result = '';
+
+        $ASCII = self::$ASCII;
+        $decompMap = self::$D;
+        $combClass = self::$cC;
+        $ulenMask = self::$ulenMask;
+        if ($c) {
+            $compatMap = self::$KD;
+        }
+
+        $c = array();
+        $i = 0;
+        $len = strlen($s);
+
+        while ($i < $len) {
+            if ($s[$i] < "\x80") {
+                // ASCII chars
+
+                if ($c) {
+                    ksort($c);
+                    $result .= implode('', $c);
+                    $c = array();
+                }
+
+                $j = 1 + strspn($s, $ASCII, $i + 1);
+                $result .= substr($s, $i, $j);
+                $i += $j;
+            } else {
```

Comment:
@nicolas-grekas This just me, I think this would be more readable if the 'if' statements had a 'continue' at the end thereby negating the need for the 'else'.